### PR TITLE
feat(schema): replicate-aware stats primitives + v3 run schema

### PR DESCRIPTION
**Pull request by dbworku**
**bench-matrix refactor — replicate-aware benchmark pipeline**

Shipped as a 4-PR stack (merge bottom-up, each branch is independently green and tree-verified):
1. #192 — **schema-core**: replicate stats primitives + v3 run schema
2. #193 — **suite-catalog**: restructure suites + catalog (pgbench split, cpu-generic retired, per-tier k/R)
3. #194 — **harness-k**: per-suite `ptsTimesToRun` (k) in the sandbox preamble
4. #195 — **results-replicate**: fold replicate shards + replicate-aware leaderboard ranking

↑ **This PR is #192 (1/4)** — based on `main`.

---

The schema foundation for the replicate-based benchmark pipeline. Pure-additive and inert in production (all current data is R=1): nothing emits v3 or calls the new stats yet — the results package upstack (#195) consumes them.

- **analysis.ts**: a hierarchical (two-level) bootstrap median interval and a difference-in-medians CI separation test (`canSeparate`), each degenerating to the ordinary percentile bootstrap at R=1 so the render path is byte-identical.
- **run.ts**: `schemaVersion "3"` (`parseRun` accepts `"2"` and `"3"`; already-published v2 Runs read unchanged, never migrated in place). Adds the optional `MetricResult.replicates` breakdown and shard-level `Run.replicateIndex`, with `narrow()` cross-field validation (≥2 distinct replicates whose pooled samples are exactly the union of the slices).

---
## Summary by cubic
Adds replicate-aware stats primitives and a v3 run schema for multi-sandbox benchmarks. R=1 output stays identical; v2 runs parse unchanged.

- **New Features**
  - `packages/schema/src/analysis.ts`: Add `hierarchicalBootstrapMedianInterval` and replicate-aware `bootstrapMedianDifferenceInterval` with exact per-replicate Mann–Whitney U (order-invariant) and `pValue`/`minAttainablePValue`/`method`. Deterministic by seed; hierarchical bootstrap; degenerates to ordinary bootstrap at R=1 and to a point when a side pools to one sample.
  - `packages/schema/src/run.ts`: Accept `schemaVersion` "2" or "3". Add `MetricResult.replicates` and shard `replicateIndex` with validation (≥2 replicates, distinct indices, finite samples, pooled union matches `samples`, `aggregates.n === samples.length`). V3-only gating rejects v2 runs with replicate fields and runs mixing `replicateIndex` with folded `replicates`; unknown versions rejected.

- **Migration**
  - No changes needed; existing v2 data reads as-is.
  - To enable replicate-aware stats, emit `schemaVersion: "3"` with per-metric `replicates` and shard `replicateIndex`. R=1 remains byte-identical.

<sup>Written for commit abee8db1cfdfb62e8541ce304396d5fa913e6da0. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/starslingdev/hpc-sandbox-benchmarks/pull/192?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Added schema version 3 support for replicate-level metric data, including shard vs aggregate replicate modeling and pooled-sample consistency checks.
  * Introduced replicate-aware hierarchical bootstrap median interval and median-difference interval with seeded, reproducible resampling and separation verdicts.
* **Bug Fixes**
  * Strengthened validation to reject incompatible schema versions and inconsistent/invalid replicate payloads (including mutual-exclusion and “lone replicate” cases).
* **Tests**
  * Expanded coverage for interval correctness (edge cases like `resamples: 0`), bracketing, reproducibility, and option/input validation; added v2/v3 boundary tests for run parsing.

### CODE REVIEW

**Comment by dbworku** [comment]
* **#195** <a href="https://app.graphite.com/github/pr/starslingdev/hpc-sandbox-benchmarks/195?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#194** <a href="https://app.graphite.com/github/pr/starslingdev/hpc-sandbox-benchmarks/194?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#193** <a href="https://app.graphite.com/github/pr/starslingdev/hpc-sandbox-benchmarks/193?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#192** <a href="https://app.graphite.com/github/pr/starslingdev/hpc-sandbox-benchmarks/192?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/starslingdev/hpc-sandbox-benchmarks/192?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>merge-ready</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/starslingdev?ref=merge-queue-instructions-in-stack-comment&prId=192).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/schema/src/analysis.ts [hunk 1]
```diff
@@ -242,6 +242,172 @@ export function bootstrapMedianInterval(
...
+	resamples: number;
+	/**
+	 * Whether the interval EXCLUDES 0 — the separation verdict. True iff the whole interval sits on one
+	 * side of 0 (`lo > 0 || hi < 0`); an interval that straddles or touches 0 is a tie (the two medians
+	 * are not distinguishable at this coverage). This is the replicate-aware replacement for the
+	 * Mann-Whitney tie decider: MW on Samples pooled across replicates treats clustered draws as
+	 * independent (anti-conservative), and MW on the R replicate medians can never reach significance at
+	 * the small R this benchmark runs — the difference interval sidesteps both by resampling the clusters.
+	 */
+	separated: boolean;
```
[the remark this answers is not in the document]
Comment (dbworku) [reply packages/schema/src/analysis.ts]: Resolved by making the contract explicit (ac27a215): `resamples: 0` now denotes the UNESTIMABLE single-Sample-per-side case, and the `separated` doc states that in that case the verdict is forced `false` regardless of the point bounds `lo === hi === difference` — zero statistical power can't justify a separation claim. The degenerate test pins exactly that (resamples 0, separated false, lo===hi===difference).

File: packages/schema/src/analysis.ts [hunk 2]
```diff
@@ -242,6 +242,172 @@ export function bootstrapMedianInterval(
...
+	const rng = seededRng(options.seed ?? "sandbox-benchmarks");
+	const diffs = new Float64Array(resamples);
+	for (let i = 0; i < resamples; i++) {
+		diffs[i] = hierarchicalResampleMedian(a, rng) - hierarchicalResampleMedian(b, rng);
+	}
+	const sorted = Array.from(diffs).sort((x, y) => x - y);
+	const tail = (1 - level) / 2;
+	const lo = percentile(sorted, tail);
+	const hi = percentile(sorted, 1 - tail);
+	return { difference, lo, hi, level, resamples, separated: lo > 0 || hi < 0 };
```
[the remark this answers is not in the document]
Comment (dbworku) [reply packages/schema/src/analysis.ts]: Documented and deferred (owner decision), because this is not a live bug: the leaderboard always calls this in a fixed (higher-ranked, lower-ranked) order, so its verdict is deterministic and byte-reproducible — the asymmetry only bites a direct caller that passes swapped args. Making it order-invariant is a real methodology change (independent per-group resampling) that would shift R>=2 interval values, so instead ac27a215 documents the fixed-order contract on the function + `separated` and adds a reversal test pinning current behavior (`difference` negates exactly; a clearly-separated pair stays separated either way). A caller needing symmetry must canonicalize the arguments.

File: packages/schema/src/analysis.ts [hunk 3]
```diff
@@ -242,6 +242,183 @@ export function bootstrapMedianInterval(
...
```
Comment (cosmic-emerald) [inline packages/schema/src/analysis.ts]: The current guard only catches one pooled sample per side, so one sandbox with several trials can still return `separated: true.` Is that ok or do we need multiple sandboxes?

File: packages/schema/src/analysis.ts [hunk 4]
```diff
@@ -242,6 +242,183 @@ export function bootstrapMedianInterval(
...
+	const difference = percentileOf(pooledA, 0.5) - percentileOf(pooledB, 0.5);
+	// Neither side carries any spread when both pool to a single Sample, so there is no basis for a
+	// separation verdict — degenerate to a point interval (resamples: 0, tied) exactly as the single-level
+	// and hierarchical median bootstraps do, rather than reporting 10 000 identical draws as real power.
+	if (pooledA.length === 1 && pooledB.length === 1) {
+		return { difference, lo: difference, hi: difference, level, resamples: 0, separated: false };
+	}
+	const rng = seededRng(options.seed ?? "sandbox-benchmarks");
+	const diffs = new Float64Array(resamples);
+	for (let i = 0; i < resamples; i++) {
```
Comment (cosmic-emerald) [inline packages/schema/src/analysis.ts:457]: How should this preserve the 95% interpretation at R=3? `[[1],[2],[3]]` versus `[[4],[5],[6]]` returns `separated: true`, but complete separation occurs in `2 / C(6,3) = 10%` of exact cluster-level null allocations. Would an exact procedure or an underpowered verdict be safer here?

File: packages/schema/src/run.ts [hunk 5]
```diff
@@ -210,22 +263,45 @@ export type ProviderRun = typeof providerRunSchema.infer;
...
 	// free-form string ("tomorrow") that silently breaks newest-first ordering.
 	generatedAt: "string.date.iso",
 	"sourceRunUrl?": "string",
+	// The replicate sandbox index a SHARD Run was measured under (the `--replicate` argument). Present on
+	// a per-replicate shard; the aggregate reads it to key {@link MetricResult.replicates} and drops it
+	// from the merged Run (which spans every replicate). Absent on legacy shards and aggregated Runs.
+	"replicateIndex?": "number.integer >= 0",
 	targetSpec: targetSpecSchema,
 	providers: providerRunSchema.array(),
+}).narrow((run, ctx) => {
```
Comment (cosmic-emerald) [inline packages/schema/src/run.ts:289]: Should a v3 Run be rejected when it contains both `replicateIndex` and metric-level `replicates`?

**Review by cosmic-emerald — APPROVED** [review]

File: packages/schema/src/analysis.ts (hunk 4 above)
Comment (dbworku) [reply packages/schema/src/analysis.ts:457]: Not sure that I understand this one. Let's whiteboard so I can understand better

File: packages/schema/src/run.ts (hunk 5 above)
Comment (dbworku) [reply packages/schema/src/run.ts:289]: Fixed in ba8e6f2 — the runSchema `.narrow` now rejects a v3 Run carrying **both** `replicateIndex` (per-replicate shard marker) and metric-level `replicates` (aggregate marker). The aggregate drops `replicateIndex` when it folds, so a Run with both is neither a clean shard nor a clean aggregate; rejecting it at the boundary keeps a consumer from having to guess which level it's reading. Added a run.test.ts case asserting the rejection.
